std/uni.d

```diff
@@ -8214,6 +8215,162 @@ unittest //12428
+{
+    static if (!is(Unqual!(ElementEncodingType!Range) : const(dchar)))
+    {
+        import std.utf.byDchar;
+
+        // Decode and try again
+        return toCaser!(indexFn, maxIdx, tableFn, Range)(str.byDchar);
+    }
+    else
+    {
+        static struct ToCaserImpl
+        {
+            this(ref Range r)
+            {
+                this.r = r;
+            }
+
+            @property bool empty()
+            {
+                return !nLeft && r.empty;
+            }
+
+            @property auto front() @safe
+            {
+                if (!nLeft)
+                {
+                    dchar c = r.front;
+                    const idx = indexFn(c);
+                    if (idx == ushort.max)
+                    {
+                        buf[0] = c;
+                        nLeft = 1;
+                    }
+                    else if (idx < maxIdx)
+                    {
+                        buf[0] = tableFn(idx);
+                        nLeft = 1;
+                    }
+                    else
+                    {
+                        auto val = tableFn(idx);
+                        // unpack length + codepoint
+                        nLeft = val >> 24;
+                        assert(nLeft <= buf.length);
+                        buf[nLeft - 1] = cast(dchar)(val & 0xFF_FFFF);
+                        foreach (j; 1 .. nLeft)
+                            buf[nLeft - j - 1] = tableFn(idx + j);
+                    }
+                }
+                return buf[nLeft - 1];
+            }
+
+            void popFront()
+            {
+                if (!nLeft)
+                    front();
+                --nLeft;
+                if (!nLeft)
+                    r.popFront();
+            }
+
+            static if (isForwardRange!Range)
+            {
+                @property auto save()
+                {
+                    auto ret = this;
+                    ret.r = r.save;
+                    return ret;
+                }
+            }
+
+          private:
+            Range r;
+            uint nLeft;
+            dchar[3] buf; // = void;
+        }
+
+        return ToCaserImpl(str);
+    }
+}
+
+/*********************
+ * Convert input range or string to upper or lower case.
+ *
+ * Does not allocate memory.
+ * Characters in UTF-8 or UTF-16 format that cannot be decoded
+ * are treated as $(XREF utf, replacementDchar).
+ *
+ * Params:
+ *      str = string or range of characters
+ *
+ * Returns:
+ *      an InputRange of dchars
+ *
+ * See_Also:
+ *      $(LREF toUpper), $(LREF toLower)
+ */
+
+auto toLowerCase(Range)(Range str)
+    if (isSomeString!Range ||
+        isInputRange!Range && isSomeChar!(ElementEncodingType!Range))
+{
+    return toCaser!LowerTriple(str);
+}
+
+/// ditto
+auto toUpperCase(Range)(Range str)
+    if (isSomeString!Range ||
+        isInputRange!Range && isSomeChar!(ElementEncodingType!Range))
```

Comment:
Delete the first clause in the conjunction:

    if (isInputRange!Range && isSomeChar!(ElementEncodingType!Range))